### DISCUSSION TITLE
Any recommendation to parse arguments that are space separated

### DISCUSSION THREAD

**Discussion by snicoll** [remkop/picocli#2342] [Q&A] [answered]
Hi everyone,

I am using picocli to build a tool that needs to compile some Java code. That part of the process can be configured with compiler options. Looking at how other tools (such as Maven) deal with this, it's either via one entry per argument or a `String` that contains the complete list of arguments.

The compiler I use requires those options to be provided one by one. After a bit of back and forth I ended up with:

```java
@Option(names = { "--compiler-arguments" }, split = " ",
		description = "Compiler arguments to use to compile generated sources.")
private List<String> compilerArguments;
```

This works well, but I am wondering if that's the best option. In particular, I wonder if the project has some sort of parsing for command line where single quote or double quote are used to capture an argument with space in its value. Ideally, I would like to support something like:

```shell
mytool --compiler-arguments --enable-preview -Dtest='My Value' -Danother=Another
```

which would parse a list with 3 items.

I am also considering adding support for a file argument and I wonder if the option I've chosen would prevent me to use that in an idiomatic fashion.

Thanks in advance!

**Comment by remkop** [comment]
One idea would be to have a two-phased approach, where in the 2nd phase, you parse the input that was gathered into the --compiler-arguments list. So, you would have one class with the `@Option(names = { "--compiler-arguments" }` -annotated field,
and another class with `@Option` annotations for the `--enable-preview`, `-D`, and other options.
Picocli will then be able to take care of the values with spaces when parsing `-Dtest='My Value'`. The user manual has an example of parsing in two phases: "[Controlling the locale](https://picocli.info/#_controlling_the_locale)".

Picocli has some [support for quoted values](https://picocli.info/#_quoted_values), for cases where the quotes were not removed by the shell.

Hope this is helpful.

**Reply by snicoll** [reply]
Thanks for the feedback.I don't want to reproduce the compiler options in my CLI. What I want really is that the user provides me the options they want for  compiling the code and I am just passing that to the compiler without extra fuss.

I think your last link is what I am looking for but I wasn't able to make it work in a test.

I have `setTrimQuotes` set to true. I've tried the following:

```
"--compiler-arguments --enable preview --parameters --release 21 --nowarn"
```

Leads to

```
picocli.CommandLine$UnmatchedArgumentException: Unknown option: '--compiler-arguments --enable preview --parameters --release 21 --nowarn'

	at picocli.CommandLine$Interpreter.validateConstraints(CommandLine.java:13668)
	at picocli.CommandLine$Interpreter.parse(CommandLine.java:13620)
	at picocli.CommandLine$Interpreter.parse(CommandLine.java:13565)
	at picocli.CommandLine$Interpreter.parse(CommandLine.java:13460)
	at picocli.CommandLine.parseArgs(CommandLine.java:1552)
```

```
--compiler-arguments "--enable preview --parameters --release 21 --nowarn"
```

Leads to 

```
picocli.CommandLine$UnmatchedArgumentException: Unknown option: '--compiler-arguments "--enable preview --parameters --release 21 --nowarn"'

	at picocli.CommandLine$Interpreter.validateConstraints(CommandLine.java:13668)
	at picocli.CommandLine$Interpreter.parse(CommandLine.java:13620)
	at picocli.CommandLine$Interpreter.parse(CommandLine.java:13565)
	at picocli.CommandLine$Interpreter.parse(CommandLine.java:13460)
	at picocli.CommandLine.parseArgs(CommandLine.java:1552)
```

And 

```
--compiler-arguments "--a-param with space" --parameters --release 21 --nowarn
```

```
picocli.CommandLine$UnmatchedArgumentException: Unknown option: '--compiler-arguments "--a-param with space" --parameters --release 21 --nowarn'

	at picocli.CommandLine$Interpreter.validateConstraints(CommandLine.java:13668)
	at picocli.CommandLine$Interpreter.parse(CommandLine.java:13620)
	at picocli.CommandLine$Interpreter.parse(CommandLine.java:13565)
	at picocli.CommandLine$Interpreter.parse(CommandLine.java:13460)
	at picocli.CommandLine.parseArgs(CommandLine.java:1552)
```

The last example is really what I am looking for. I would like the split to igore the space within the quoted value. Unfortunately, I can't go as far as it fails before that.

I don't understand what I am doing wrong.

**Comment by remkop** [comment]


**Reply by remkop** [reply]
@snicoll I am not sure why your test has an error "unknown option --compiler-arguments", since you have that option defined in your application...

However, this may be helpful:
I created this test (these all pass): https://github.com/remkop/picocli/blob/main/src/test/java/picocli/Issue2342.java

I think it accomplishes your use case, can you take a look?

Note that I had to set `arity = "*"` on  `@Option(names = { "--compiler-arguments" }...` for this option to greedily consume all the remaining arguments.
The example also illustrates what `setTrimQuotes` does.

**Reply by snicoll** [reply]
> Note that I had to set arity = "*" on @Option(names = { "--compiler-arguments" }... for this option to greedily consume all the remaining arguments.

I don't want that is the point. I've removed the arity and all those tests are failing for me. Perhaps I am not explaining what I want clearly. Thanks for the feedback so far as it revealed another problem if a compiler argument value matches another option of the command (oh well).

I think my ask is as follows:

* An option has a single value. That single value has values that are space separated that I would like to be bound to `List<String>`
* Those values should not be interpreted. I understand that I need to provide some sort of information to the parser for that to happen, but I don't know what that is. I've tried to put quotes in various places but none worked.
* Assuming the above works, I am then trying to figure out what should happen if an individual value within that list has a space itself. That's mostly what we've been talking about here but it turns out that's not my biggest problem. I was hoping that this can be combined with a file where each item could be provided on its own line. Or nested quote. Again, I am not sure what Picocli does support.

I've updated the tests in https://github.com/snicoll/picocli/commit/119baa9ea5aad69aae969e0820a5d373b5be20e5

I am also happy to bound to a `String` and then do the space split myself. I actually considered that first and I was wondering if picocli has a utility where I would give it a `String` and it would give me back a `List<String>` where I can provide the split. If picocli has advanced features such as support for nested quotes, I guess that utility should exist somewhere...

**Comment by remkop** [comment]
Thanks for updating the test. I understand a bit better now (I hope). This is my understanding now: 

* Your application has two options: `--compiler-arguments` and `--target`.
* The last occurrence of `--target` should be followed by an option parameter, and that value should be assigned to the `target` field.
* All values between `--compiler-arguments` and the last occurrence of `--target` should be gathered in the `compilerArguments` list. Even if such values "look like" options (start with `--` or `-`), these values should simply be added to the `compilerArguments` list and should not be interpreted as options, therefore should not cause any "unknown option" errors.

Your requirements are unusual but doable. :-)
I have provided an example solution in the test class, using [Custom Parameter Processing](https://picocli.info/#_custom_parameter_processing).
The solution also handles quoted values, but only values that have "balanced quotes": values that start and end with the same quote character.

Please take a look at the updated tests with the custom parameter processing: https://github.com/remkop/picocli/blob/main/src/test/java/picocli/Issue2342.java

I added an `@Igore` annotation with some comment to the test cases where I don't yet understand your intention.

**Comment by remkop** [comment]
Additionally, you mentioned that

> I was hoping that this can be combined with a file where each item could be provided on its own line. Or nested quote. Again, I am not sure what Picocli does support.

Picocli has support for [@-files](https://picocli.info/#AtFiles). I believe this may be what you are looking for, please take a look.

**Comment by snicoll** [comment]
> The last occurrence of --target should be followed by an option parameter, and that value should be assigned to the target field.

I think that's making things way more complicated than I want them to be. I think the problem is that the tests are using a String[] so they make the problem a bit more artificial. 

Thanks for the help thus far but I don't think we're getting at the problem. 

The "only" thing I want is to be able to express that a value for an option can be quoted in such a way that it's not interpreted. If picocli can then split things within that single value and handle more feature like nested quotes, then it's even better but that's secondary.

I read your ignore message and I am afraid I don't understand them.

Let's assume for the sake of the examples below that a single quote is to be used to group things into a singe value. 

These examples are identical and should assign `my-file.jar` to the target option and `--target 21 -nowarn` to the compiler arguments. 

```
command --compiler-arguments '--target 21 -nowarn' --target my-file.jar
command --target my-file.jar --compiler-arguments '--target 21 -nowarn'
command --target my-file.jar --compiler-arguments --target 21 -nowarn
```

Again I totally understand that the user has to quote something somewhere to make sure that the value is interpreted correctly. For instance, this application has also a single parameter, let's call that `application.jar`. All these should lead to the same outcome (same as before + assigning `application.jar` to the single parameter).

```
command --compiler-arguments '--target 21 -nowarn' --target my-file.jar application.jar
command --target my-file.jar --compiler-arguments '--target 21 -nowarn' application.jar
command application.jar --compiler-arguments '--target 21 -nowarn' --target my-file.jar
command application.jar --target my-file.jar --compiler-arguments '--target 21 -nowarn'
command application.jar --target my-file.jar --compiler-arguments --target 21 -nowarn
```

I understand that the third previous example won't cut it with `application.jar` at the end, as I need to tell picocli that the value of the compiler arguments has ended. Does what I am describing makes even any sense? To me it really boils down to an option having a value with spaces in them.

> Picocli has support for [@-files](https://picocli.info/#AtFiles). I believe this may be what you are looking for, please take a look.

I did and I used it in some of my tests. What I was trying to explain is that I was hoping that maybe using a file could help those problems by putting each value in their own line or something. Doing so would mean that we don't need to care about quoting and a newline could be a character to split the values. Unfortunately, I don't believe this feature exists at the moment.

**Comment by remkop** [comment, marked as answer]
Okay, I think there is some confusion because of the unit tests: if the values are quoted correctly, the single-arity option `--compiler-arguments` does exactly what you need. 

In the unit tests, however, there should be no quotes because they are interpreted and removed by the shell (Bash for example).

Example code:

```java
import picocli.CommandLine;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

import java.util.Arrays;
import java.util.List;

public class MyCommand implements Runnable {
    @Option(names = { "--compiler-arguments" }, split = " ",
            description = "Compiler arguments to use to compile generated sources.")
    private List<String> compilerArguments;

    @Option(names = "--target", description = "Test for presence in compiler arguments")
    private String target;

    @Parameters(description = "optional jar file", arity = "0..1")
    private String jarFile;

    public static void main(String[] args) {
        System.out.printf("%s (%d elements)%n", Arrays.asList(args), args.length);
        System.out.print("String[] args = new String[] {");
        Arrays.asList(args).forEach(s -> System.out.printf("\"%s\", ", s));
        System.out.println("}; // for unit test");

        new CommandLine(new MyCommand())
                .setAllowOptionsAsOptionParameters(true)
                .execute(args);
    }

    @Override
    public void run() {
        System.out.printf("CompilerArguments: %s%n", compilerArguments);
        System.out.printf("Target:            %s%n", target);
        System.out.printf("jarFile:           %s%n", jarFile);
    }
}
```

In git-bash, I created an alias so I can run this with `command`:

```
$ alias command="~/pkg/jdk/bin/java -cp '.;~/picocli-4.7.6.jar' MyCommand"
```

Now, going through your examples in git-bash:

This works exactly as expected. Note that quotes are used by the shell to pass a single value to `--compiler-arguments`, and the Java application does not see those quotes.
```
$ command --compiler-arguments '--target 21 -nowarn' --target my-file.jar
[--compiler-arguments, --target 21 -nowarn, --target, my-file.jar] (4 elements)
String[] args = new String[] {"--compiler-arguments", "--target 21 -nowarn", "--target", "my-file.jar", }; // for unit test
CompilerArguments: [--target, 21, -nowarn]
Target:            my-file.jar
jarFile:           null
```

Same here: quoted space-separated values passed on the git-bash shell are received as a single unquoted value by the Java application
```
$ command --target my-file.jar --compiler-arguments '--target 21 -nowarn'
[--target, my-file.jar, --compiler-arguments, --target 21 -nowarn] (4 elements)
String[] args = new String[] {"--target", "my-file.jar", "--compiler-arguments", "--target 21 -nowarn", }; // for unit test
CompilerArguments: [--target, 21, -nowarn]
Target:            my-file.jar
jarFile:           null
```

Without quotes, however, the values are passed as separate parameters and picocli takes the first value to be the `--compiler-arguments` parameter, and has trouble parsing the remainder...
```
$ alias command2="~/pkg/jdk/bin/java -cp '.;picocli-4.7.6.jar' -Dpicocli.trace=INFO MyCommand"

$ command2 --target my-file.jar --compiler-arguments --target 21 -nowarn
[--target, my-file.jar, --compiler-arguments, --target, 21, -nowarn] (6 elements)
String[] args = new String[] {"--target", "my-file.jar", "--compiler-arguments", "--target", "21", "-nowarn", }; // for unit test
[picocli INFO] Picocli version: 4.7.6, JVM: 17.0.8 (Oracle Corporation Java HotSpot(TM) 64-Bit Server VM 17.0.8+9-LTS-211), OS: Windows 10 10.0 amd64
[picocli INFO] Parsing 6 command line args [--target, my-file.jar, --compiler-arguments, --target, 21, -nowarn]
[picocli INFO] Setting field String MyCommand.target to 'my-file.jar' (was 'null') for option --target on MyCommand@6bdf28bb
[picocli INFO] Adding [--target] to field java.util.List<String> MyCommand.compilerArguments for option --compiler-arguments on MyCommand@6bdf28bb
[picocli INFO] Setting field String MyCommand.jarFile to '21' (was 'null') for args[0+ (0)] at position 0 on MyCommand@6bdf28bb
Unknown option: '-nowarn'
Usage: <main class> [--target=<target>]
                    [--compiler-arguments=<compilerArguments>]... [<jarFile>]
      [<jarFile>]         optional jar file
      --compiler-arguments=<compilerArguments>
                          Compiler arguments to use to compile generated
                            sources.
      --target=<target>   Test for presence in compiler arguments

```

With positional parameter. Again, when compiler args are quoted, it works as expected:
```
$ command --compiler-arguments '--target 21 -nowarn' --target my-file.jar application.jar
[--compiler-arguments, --target 21 -nowarn, --target, my-file.jar, application.jar] (5 elements)
String[] args = new String[] {"--compiler-arguments", "--target 21 -nowarn", "--target", "my-file.jar", "application.jar", }; // for unit test
CompilerArguments: [--target, 21, -nowarn]
Target:            my-file.jar
jarFile:           application.jar
```

With positional parameter. Again, when compiler args are quoted, it works as expected:
```
$ command --target my-file.jar --compiler-arguments '--target 21 -nowarn' application.jar
[--target, my-file.jar, --compiler-arguments, --target 21 -nowarn, application.jar] (5 elements)
String[] args = new String[] {"--target", "my-file.jar", "--compiler-arguments", "--target 21 -nowarn", "application.jar", }; // for unit test
CompilerArguments: [--target, 21, -nowarn]
Target:            my-file.jar
jarFile:           application.jar
```

With positional parameter. Again, when compiler args are quoted, it works as expected:
```
$ command application.jar --compiler-arguments '--target 21 -nowarn' --target my-file.jar
[application.jar, --compiler-arguments, --target 21 -nowarn, --target, my-file.jar] (5 elements)
String[] args = new String[] {"application.jar", "--compiler-arguments", "--target 21 -nowarn", "--target", "my-file.jar", }; // for unit test
CompilerArguments: [--target, 21, -nowarn]
Target:            my-file.jar
jarFile:           application.jar
```

With positional parameter. Again, when compiler args are quoted, it works as expected:
```
$ command application.jar --target my-file.jar --compiler-arguments '--target 21 -nowarn'
[application.jar, --target, my-file.jar, --compiler-arguments, --target 21 -nowarn] (5 elements)
String[] args = new String[] {"application.jar", "--target", "my-file.jar", "--compiler-arguments", "--target 21 -nowarn", }; // for unit test
CompilerArguments: [--target, 21, -nowarn]
Target:            my-file.jar
jarFile:           application.jar
```

With positional parameter. Without the quotes, picocli does not know where to assign the '21' and '-nowarn' values to.
```
$ command application.jar --target my-file.jar --compiler-arguments --target 21 -nowarn
[application.jar, --target, my-file.jar, --compiler-arguments, --target, 21, -nowarn] (7 elements)
String[] args = new String[] {"application.jar", "--target", "my-file.jar", "--compiler-arguments", "--target", "21", "-nowarn", }; // for unit test
Unmatched arguments from index 5: '21', '-nowarn'
Usage: <main class> [--target=<target>]
                    [--compiler-arguments=<compilerArguments>]... [<jarFile>]
      [<jarFile>]         optional jar file
      --compiler-arguments=<compilerArguments>
                          Compiler arguments to use to compile generated
                            sources.
      --target=<target>   Test for presence in compiler arguments
```

**Reply by snicoll** [reply]
> In the unit tests, however, there should be no quotes because they are interpreted and removed by the shell (Bash for example).

I think that's what tripped me. Thanks, I confirm this works as expected.

**Comment by remkop** [comment]
Note that different shells have different quoting rules/support:

In the Windows Command Prompt shell, single quotes are not recognized and are passed on to the Java application.

```
%JAVA_HOME%\bin\java -cp .;picocli-4.7.6.jar MyCommand --compiler-arguments '--target 21 -nowarn' --target my-file.jar
[--compiler-arguments, '--target, 21, -nowarn', --target, my-file.jar] (6 elements)
String[] args = new String[] {"--compiler-arguments", "'--target", "21", "-nowarn'", "--target", "my-file.jar", }; // for unit test
Unknown option: '-nowarn''
Usage: <main class> [--target=<target>]
                    [--compiler-arguments=<compilerArguments>]... [<jarFile>]
      [<jarFile>]         optional jar file
      --compiler-arguments=<compilerArguments>
                          Compiler arguments to use to compile generated
                            sources.
      --target=<target>   Test for presence in compiler arguments
```

However, double quotes work as expected in Windows Command Prompt shell:

```
%JAVA_HOME%\bin\java -cp .;picocli-4.7.6.jar MyCommand --compiler-arguments "--target 21 -nowarn" --target my-file.jar
[--compiler-arguments, --target 21 -nowarn, --target, my-file.jar] (4 elements)
String[] args = new String[] {"--compiler-arguments", "--target 21 -nowarn", "--target", "my-file.jar", }; // for unit test
CompilerArguments: [--target, 21, -nowarn]
Target:            my-file.jar
jarFile:           null
```
